perf(ivy): removes generation of comments

divdavem opened the pull request:
## PR Type
What kind of change does this PR introduce?

This PR reduces the time needed to insert elements in the DOM for some use cases such as the one from the [tree_perf benchmark](https://github.com/angular/angular/blob/master/modules/benchmarks/e2e_test/tree_perf.ts).

## What is the current behavior?

An empty HTML comment is inserted in the DOM for each container node (`LContainerNode`) to mark the position where it is, and this is used when calling `insertBefore`.

## What is the new behavior?

Comments are no longer inserted in the DOM. The tree of logical nodes is used to look for the next element, to be passed to the `insertBefore` function. This reduces the time needed to insert elements, as it is less costly to explore the tree of `LNode` elements rather than adding comments in the DOM.

## Does this PR introduce a breaking change?

Yes: queries for `ElementRef` on `ng-template` elements no longer return a pointer to the generated comments, as those comments are no longer generated (cf the test I changed). I discussed a bit with @pkozlowski-opensource, it is debatable whether this edge case should be supported.

## Other information

I will add some tests to more thoroughly check that everything works fine after this change. (Especially, I did not fully consider the case of projected nodes).

As suggested by @marclaval, we could add some optimizations to avoid re-computing the next element when we know it is the same (for example: when adding several views at the end of the same container).

mhevery commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -47,24 +47,56 @@ export function findNativeParent(containerNode: LContainerNode): RNode|null {
+    child = (node as LViewNode).child;
+  }
+  while (child) {
+    const native = findFirstNativeNode(child);
```
This is a recursive function. Recursive functions can't be inlined by JIT and as a result will run slower than non recursive function. Given that this is a hot path, we should solve this in a way which is non-recursive.

mhevery commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -47,24 +47,56 @@ export function findNativeParent(containerNode: LContainerNode): RNode|null {
+  if (parent) {
+    const parentType = parent.flags & LNodeFlags.TYPE_MASK;
+    if (parentType === LNodeFlags.Container || parentType === LNodeFlags.View) {
+      return findBeforeNode(parent);
```
This is also recursive.

divdavem commented:
@mhevery Thank you for your comments. I have removed the recursion.
There is now a stack in the `findFirstNativeNode`, which is not very nice, but I am thinking about removing it and storing the first native node in the `LNode` structure instead (but we need to update it when it changes).

mhevery commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -218,8 +248,15 @@ export function insertView(
+    if (!beforeNode) {
+      let containerNextNativeNode = container.data.nextNative;
+      if (containerNextNativeNode === undefined) {
+        containerNextNativeNode = container.data.nextNative = findBeforeNode(container, null);
```
I don't think I understand the value of saving the data here. What would happen if we did not save it?

divdavem replied:
It is not mandatory to save the value here. This is an optimization suggested by @marclaval: when we need to add many views at the end of the same container, it avoids re-computing multiple times the next node.

mhevery commented on packages/core/src/render3/interfaces/container.ts:
```diff
@@ -66,6 +68,15 @@ export interface LContainer {
+   * When it is undefined, it means the value has not been computed yet.
+   * Otherwise, it contains the result of findBeforeNode(container, null).
+   */
+  nextNative: RNode|null|undefined;
```
I have very mixed feelings about this field. This field will always be `undefined` (unless we are in the process of inserting) hence it contains transient data. I don't fully understand the need for this so it is hard for me to suggest a different solution, but could we solve this without storing any transient data in the data structure?

divdavem replied:
Indeed, this is transient data, and it is not mandatory to store it (just a cache to avoid re-computing the value). If we want to store it elsewhere, we could instead have a global stack (but it is not perfect either). It is the usual space-time trade-off.

kara replied:
Here are two alternatives:

1) We can create a global in `node_manipulation` called `containerNextNative` that is set in `insertView` and cleared out in `cr()` (through a reset function maybe). 

2) We could also store this where the comment node was previously stored - `LContainerNode.native`. It serves the same purpose as the comment did, so conceptually I think it's fine. 

Either way, we'd save 8 bytes per container instance.

mhevery commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -122,6 +121,11 @@ export interface LNode {
+  /**
+   * If this node is projected, pointer to the corresponding projection node.
+   */
+  projectionNode: LProjectionNode|null;
```
Why do we need this new field? This was not here before and the system was functioning.

My concern is that I would like to keep the `LNode` data structure as small as possible.

divdavem replied:
Before adding this new field, my PR was not correctly working for projected nodes.
See the test I added in `integration_spec.ts`: when adding a view to a container which is projected, we have no way without this field to find the node before which the new view should be inserted, as it is in the shadow DOM, and the `parent` and `next` field contain references to nodes in the light DOM only (those references should not be followed in case a node is projected because different selectors can project those nodes at different places).
See also #21730 which underlines a problem with projected nodes with the current implementation.

divdavem commented:
@mhevery Thank you for your comments!
I will do some more changes in this PR (rebasing and solving conflicts, adding full support for projected nodes, including fixing #21730).

divdavem commented:
I have now made the changes I planned to do in this PR.

Instead of storing the link to the parent projection node (which then required looping on the array in the projection node data to find the current node), I have now changed the implementation to store the next projected node in the same parent projection node, and at the end of this linked list, the projection node itself, because it is in this order that `findFirstNativeNode`, `addRemoveViewFromContainer` and `findBeforeNode` need to traverse the nodes.

Now that projected nodes are linked together, it is a linked list, so I have also removed the array from the data associated to a projection node to just have a reference to the first and last node that were previously stored in an array. Even if nodes can be projected multiple times, a re-projection cannot re-order or remove individual nodes from a previous projection, it can only include the full previous linked list, with potential extra nodes before and after the included list. So as I now store the first and last node of the linked list I do not loose information. Anyway, I am not sure if it is really useful to keep information about any previous re-projection apart from the final one.

@mhevery What do you think about these changes? Do you think I have to do any other changes before this PR is integrated? Thank you for your review!

divdavem commented:
After rebasing, it seems that a new test is failing. I will investigate...

divdavem commented:
It was a mistake while rebasing... it is now fixed!

divdavem committed:
perf(ivy): removes generation of comments

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -6,65 +6,135 @@
+ * @param rootNode The root node at which the lookup should stop.
+ * @return The following node in the LNode tree.
+ */
+function findFollowingNode(initialNode: LNode, rootNode: LNode): LNode|null {
```
Nit: `findFollowingNode` and `getNextNode` sound pretty similar. Any way to capture in the name that this will search up the tree if a next node isn't found?

`getNextOrParentNode`?

divdavem replied:
Indeed, the difference was not very clear. I have called it `getNextOrParentSiblingNode` (as it returns the next sibling of a parent, and not the parent itself, even though the name starts being a bit long...).

divdavem committed:
refactor(ivy): code review changes

divdavem commented on packages/core/src/render3/instructions.ts:
```diff
@@ -989,11 +989,14 @@ export function container(
   // If the direct parent of the container is a view, its views (including its comment)
```
Ok

divdavem commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1044,7 +1048,11 @@ export function containerRefreshStart(index: number): void {
+  container.data.nextIndex = 0;
+  ngDevMode &&
+      assertEqual(
+          container.data.nextNative === undefined, true, 'container.data.nextNative === undefined');
```
Ok

divdavem commented on packages/core/src/render3/interfaces/projection.ts:
```diff
@@ -9,14 +9,13 @@
-export type LProjection = Array<LElementNode|LTextNode|LContainerNode>;
-
+export interface LProjection {
+  first: LElementNode|LTextNode|LContainerNode|null;
```
Ok. I have renamed the fields `head` and `tail` in `LProjection`. I have not changed `LView`/`LContainer` not to add too many changes in this PR (and just using the rename function of Visual Studio Code did not seem to be enough). This could be done later.

divdavem commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -6,65 +6,135 @@
+  const pNextOrParent = node.pNextOrParent;
+  if (pNextOrParent) {
+    const type = pNextOrParent.flags & LNodeFlags.TYPE_MASK;
+    return type === LNodeFlags.Projection ? null : pNextOrParent;
```
Ok

divdavem commented:
@kara Thank you very much for your comments. I have taken them into account. Please tell me if there is anything more to change.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -218,7 +218,8 @@ export function createLNode(
     query: query,
-    tNode: null
+    tNode: null,
+    pNextOrParent: null
```
Why is this called `pNextOrParent` rather than just `pNext`? When is it a parent?

divdavem replied:
Thank you for your review and for integrating my PR!

To answer your question, `pNextOrParent` is a parent at the end of the linked list (as documented [here](https://github.com/angular/angular/blob/5bd93b1f0f904532e74d0992114c16118aef30a0/packages/core/src/render3/interfaces/node.ts#L126)). In this case, it contains the projection node itself, which can easily be recognized as the parent because of its type, because projection nodes are never nested. In case of re-projection, instead of adding the projection node as a child of another one, all the children of the nested projection node are directly included in the parent projection node, as implemented here: https://github.com/angular/angular/blob/5bd93b1f0f904532e74d0992114c16118aef30a0/packages/core/src/render3/instructions.ts#L1306

So if `pNode` is a projection node that is not re-projected and contains a non-empty list of projected nodes, then `pNode.data.tail.pNextOrParent` is `pNode` itself. This is necessary to be able to easily find the first native node following a projected node.

mhevery commented:
http://test/OCL:183458846:BASE:183458968:1517010872186:e0bb7cce
